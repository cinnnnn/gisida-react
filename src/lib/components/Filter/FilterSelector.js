import React, { Component }from 'react';
import { connect } from 'react-redux';
import { catchZeroCountClicks } from '../../utils';
import PropTypes from 'prop-types';

const mapStateToProps = (state, ownProps) => {
  return {
  }
}


class FilterSelector extends Component {
  constructor(props) {
    super(props);
    this.state = Object.assign({}, this.props.filter, {
      isLinux: (window.navigator.platform.indexOf('Linux') !== -1),
    });
  }

  componentWillReceiveProps(nextProps) {
    this.setState(nextProps.filter);
  }

  onSearchClear = (e, filterKey) => {
    this.searchEl.value = '';
    this.props.searchFilterOptions(e, filterKey);
  }

  render() {
    const { options, isLinux } = this.state;
    const { filterKey, toggleAllOn } = this.props;
    if (!options || !(Object.keys(options)).length) {
      return null;
    }

    const selectedOptions = [];
    const filterOptions = [];
    const inactiveOptions = [];
    const optionKeys = Object.keys(options);
    let option;
    let optionEl;
    const totals = [];

    for (let i = 0; i < optionKeys.length; i += 1) {
      option = options[optionKeys[i]];
      optionEl = (
        <li key={i} className={`optionItem${option.count ? '' : ' inactive'}`}>
          <input
            type="checkbox"
            id={optionKeys[i]}
            value={optionKeys[i]}
            checked={option.enabled}
            onChange={(e) => { this.props.onFilterOptionClick(e, filterKey); }}
          />
          <label
            htmlFor={optionKeys[i]}
            key={optionKeys[i]}
            data-count={option.count}
            className={`optionLabel${option.enabled ? ' enabled' : ''}`}
            onClick={(e) => { this.catchZeroCountClicks(e); }}
            role="presentation"
            tabIndex="-1"
          >
            <span
              className={`filter-option-label${isLinux ? ' linux' : ''}`}
              title={optionKeys[i]}
            >
              {optionKeys[i]}
            </span>
            <span>{option.count}</span>
          </label>
        </li>
      );
      if (option.count && !option.hidden) {
        totals.push(option.count);
        filterOptions.push(optionEl);
      }
    }

    const totalOptions = totals.reduce((a, b) => a + b, 0);

    return (
      <div>
        <div className="search-column">
          <span className="searchBtn" role="button">
            <span className="glyphicon glyphicon-search" />
          </span>
          <input
            ref={(el) => { this.searchEl = el; }}
            type="text"
            className="filterSearch"
            placeholder="Search Options"
            onChange={(e) => { this.props.searchFilterOptions(e, filterKey); }}
          />
          {this.searchEl && this.searchEl.value !== '' ? (
            <span
              className="clearSearch"
              role="button"
              onClick={(e) => { this.onSearchClear(e, filterKey); }}
              tabIndex="0"
            >
              <span className="glyphicon glyphicon-remove" />
            </span>
          ) : ''}
        </div>
        <ul id="filter-group" className="filterGroup" key={filterKey}>
          <li>
            {totalOptions ? (
              <div>
                <input
                  type="checkbox"
                  id={`all-${filterKey}`}
                  checked={!toggleAllOn}
                  onChange={(e) => { this.props.onToggleAllOptions(e, toggleAllOn, filterKey); }}
                />
                <label
                  className="optionLabel"
                  htmlFor={`all-${filterKey}`}
                >
                  <span className={`filter-option-label${isLinux ? ' linux' : ''}`}>(All)</span>
                  <span>({totalOptions})</span>
                </label>
              </div>
            ) : (
              <span className="noOptions">(No options available)</span>
            )}
          </li>
          {selectedOptions.concat(filterOptions, inactiveOptions)}
        </ul>
      </div>
    );
  }
}

FilterSelector.propTypes = {
  filter: PropTypes.objectOf(PropTypes.any).isRequired,
  onFilterOptionClick: PropTypes.func.isRequired,
  searchFilterOptions: PropTypes.func.isRequired,
  onToggleAllOptions: PropTypes.func.isRequired,
  toggleAllOn: PropTypes.bool.isRequired,
  filterKey: PropTypes.string.isRequired,
};

export default connect(mapStateToProps)(FilterSelector);
