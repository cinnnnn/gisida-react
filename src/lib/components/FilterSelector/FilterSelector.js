import React from 'react';
import { connect } from 'react-redux';
import PropTypes from 'prop-types';
import { Actions, prepareLayer  } from 'gisida';

require('./FilterSelector.scss');

const mapStateToProps = (state, ownProps) => {
  const layers = state.MAP.layers;
  let layerObj;
  // Get visible layer 
  Object.keys(layers).forEach((key) => {
    const layer = layers[key];
    if (layer.filterOptions && layer.visible) {
      layerObj = layer;
    }
  });

  // Set layer to undefined of layer is fro diffrent region
  const currentRegion = state.REGIONS.filter(region => region.current)[0];
  layerObj = (layerObj && layerObj.region === currentRegion.name) ? layerObj : undefined;
  return {
    layerObj: layerObj
  }
}

class FilterSelector extends React.Component {

  filterData = (filterOptions) => {
    const dispatch = this.props.dispatch;
    const layer = this.props.layerObj;
    window.GisidaMap.removeLayer(layer.id);
    window.GisidaMap.removeSource(layer.id);
    // // dispach prepare layer if layer data  has not been loaded into props
      prepareLayer(layer, dispatch, filterOptions);
  }

  componentWillReceiveProps(nextProps) {
    if (nextProps.layerObj && nextProps.layerObj.filterOptions) {
      let filterOptions = nextProps.layerObj.filterOptions.map((opt) => {
        const optVal = {};
        optVal[opt] = true;
        return optVal;
      });
      if (this.state && this.state.options) {
        filterOptions = filterOptions.concat([this.state.options]);
      }
      this.setState({ options: Object.assign.apply(this, filterOptions) });
    } else this.setState({ options: {} });
  }

  updateOptions(val) {
    const options = this.state.options;
    options[val] = !options[val];
    this.setState({ options });
    this.filterData(options);
  }

  render() {
    if (this.state && this.state.options) {
      return (
        <nav id="filter-group" className="filter-group">
          {Object.keys(this.state.options).map(val =>
            (<span key={`label_${val}`}>
              <input
                type="checkbox"
                id={val}
                key={`input_${val}`}
                value={val}
                onChange={(e) => {
                  this.updateOptions(e.target.value);
                }}
                checked={this.state.options[val]}
              />
              <label htmlFor={val} key={`label_${val}`}>
                {val}
              </label>
            </span>))
          }
        </nav>
      );
    }
    return (<nav />);
  }
}

FilterSelector.propTypes = {
  layerObj: PropTypes.objectOf(PropTypes.any),
};

FilterSelector.defaultProps = {
  layerObj: {},
};


export default connect(mapStateToProps)(FilterSelector);
