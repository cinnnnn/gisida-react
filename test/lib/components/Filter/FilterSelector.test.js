import React from 'react';
import { FilterSelector }  from '../../../../src/lib/components/Filter/FilterSelector.js'
import { shallow } from 'enzyme';
import toJson from 'enzyme-to-json';

const onFilterOptionClick = jest.fn();
const searchFilterOptions = jest.fn();
const onToggleAllOptions = jest.fn();

const componentWrapper = shallow(
	<FilterSelector 
		filter={{}}
		onFilterOptionClick={onFilterOptionClick}
		searchFilterOptions={searchFilterOptions}
		onToggleAllOptions={onToggleAllOptions}
		toggleAllOn={false}
		filterKey=''
	/>
);

describe('FilterSelector', () => {
	it('FilterSelector component renders correctly', () => {
		const json = toJson(componentWrapper)
		expect(json).toMatchSnapshot();
	})
});