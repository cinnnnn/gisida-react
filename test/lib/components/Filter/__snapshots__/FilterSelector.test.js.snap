// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`FilterSelector FilterSelector component renders correctly 1`] = `""`;
